[UI-side compositing] Avoid sync IPC with the GPU Process when there are no buffers to swap

smfr opened the pull request:
#### 8e6829b4efded22b1546b11a89b7c3282dd103cd
<pre>
[UI-side compositing] Avoid sync IPC with the GPU Process when there are no buffers to swap
<a href="https://bugs.webkit.org/show_bug.cgi?id=254119">https://bugs.webkit.org/show_bug.cgi?id=254119</a>
rdar://106900111

Reviewed by Cameron McCormack.

When a layer is painted for the first time, we enter RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay()
with a RemoteLayerBackingStore that has no buffers. Currently, we still send sync IPC to the GPU process to swap buffers, which does
nothing in this case.

So take backing stores with no existing buffers out of the list that&apos;s sent to the GPU process, and don&apos;t message the GPU
process at all if no backing stores need swapping. For these backing stores, call `applySwappedBuffers()` with null arguments,
which will allocate the front buffer.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::hasNoBuffers const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::operator&lt;&lt;):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay):

Canonical link: <a href="https://commits.webkit.org/261838@main">https://commits.webkit.org/261838@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ff4585ca41e137207e5cf2bc3c6c2ec406d3333

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113014 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4786 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121499 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23520 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13334 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6007 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118801 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100731 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106105 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99433 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1267 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46500 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14461 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1309 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15171 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10648 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20476 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53301 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8264 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17020 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
```diff
@@ -75,14 +82,21 @@
 
-    auto& remoteRenderingBackend = layerTreeContext().ensureRemoteRenderingBackendProxy();
-    auto swapResult = remoteRenderingBackend.prepareBuffersForDisplay(WTFMove(prepareBuffersData));
+    if (prepareBuffersData.size()) {
```
`!prepareBuffersData.isEmpty()`?

smfr committed:
[UI-side compositing] Avoid sync IPC with the GPU Process when there are no buffers to swap
https://bugs.webkit.org/show_bug.cgi?id=254119
rdar://106900111

Reviewed by Cameron McCormack.

When a layer is painted for the first time, we enter RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay()
with a RemoteLayerBackingStore that has no buffers. Currently, we still send sync IPC to the GPU process to swap buffers, which does
nothing in this case.

So take backing stores with no existing buffers out of the list that's sent to the GPU process, and don't message the GPU
process at all if no backing stores need swapping. For these backing stores, call `applySwappedBuffers()` with null arguments,
which will allocate the front buffer.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::hasNoBuffers const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::operator<<):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay):

Canonical link: https://commits.webkit.org/261838@main
